src/openrct2-ui/windows/RideConstruction.cpp

```diff
@@ -233,2815 +180,2812 @@ static void CloseRideWindowForConstruction(RideId rideId)
-    track_type_t type = _currentTrackPieceType;
-    auto newCoords = GetTrackElementOriginAndApplyChanges(
-        { _currentTrackBegin, static_cast<Direction>(direction & 3) }, type, 0, &tileElement, 0);
-    if (!newCoords.has_value())
-    {
-        window_ride_construction_update_active_elements();
-        return;
-    }
-
-    // Get the previous track element to go to after the selected track element is deleted
-    inputElement.x = newCoords->x;
-    inputElement.y = newCoords->y;
-    inputElement.element = tileElement;
-    if (track_block_get_previous({ *newCoords, tileElement }, &trackBeginEnd))
-    {
-        *newCoords = { trackBeginEnd.begin_x, trackBeginEnd.begin_y, trackBeginEnd.begin_z };
-        direction = trackBeginEnd.begin_direction;
-        type = trackBeginEnd.begin_element->AsTrack()->GetTrackType();
-        gGotoStartPlacementMode = false;
-    }
-    else if (track_block_get_next(&inputElement, &outputElement, &newCoords->z, &direction))
-    {
-        newCoords->x = outputElement.x;
-        newCoords->y = outputElement.y;
-        direction = outputElement.element->GetDirection();
-        type = outputElement.element->AsTrack()->GetTrackType();
-        gGotoStartPlacementMode = false;
-    }
-    else
-    {
-        direction = _currentTrackPieceDirection;
-        type = _currentTrackPieceType;
-        newCoords = GetTrackElementOriginAndApplyChanges(
-            { _currentTrackBegin, static_cast<Direction>(direction & 3) }, type, 0, &tileElement, 0);
+        int32_t rideType = RideGetAlternativeType(currentRide);
 
-        if (!newCoords.has_value())
+        hold_down_widgets = 0;
+        if (GetRideTypeDescriptor(rideType).HasFlag(RIDE_TYPE_FLAG_IS_SHOP) || !currentRide->HasStation())
         {
-            window_ride_construction_update_active_elements();
-            return;
+            window_ride_construction_widgets[WIDX_ENTRANCE_EXIT_GROUPBOX].type = WindowWidgetType::Empty;
+            window_ride_construction_widgets[WIDX_ENTRANCE].type = WindowWidgetType::Empty;
+            window_ride_construction_widgets[WIDX_EXIT].type = WindowWidgetType::Empty;
         }
-
-        const auto& ted = GetTrackElementDescriptor(tileElement->AsTrack()->GetTrackType());
-        const rct_preview_track* trackBlock = ted.Block;
-        newCoords->z = (tileElement->GetBaseZ()) - trackBlock->z;
-        gGotoStartPlacementMode = true;
-
-        // When flat rides are deleted, the window should be reset so the ride can be placed again.
-        const auto rideId = w->rideId;
-        auto ride = get_ride(rideId);
-        const auto& rtd = ride->GetRideTypeDescriptor();
-        if (rtd.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE) && !rtd.HasFlag(RIDE_TYPE_FLAG_IS_SHOP))
+        else
         {
-            ride_initialise_construction_window(ride);
+            window_ride_construction_widgets[WIDX_ENTRANCE_EXIT_GROUPBOX].type = WindowWidgetType::Groupbox;
```

Comment:
could you go through and change all references of `window_ride_construction_widgets` to `widgets` so that you get it through the windows widgets rather than the global variable that stores the widgets. (obviously dont change the reference in OnOpen but every other one)